include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -228,71 +228,425 @@ struct __merge_sort_leaf_submitter<__internal::__optional_kernel_name<_LeafSortN
+                _BaseDiagonalsSPStorage __base_diagonals_sp_global_ptr)
+    {
+        //   |                  subrange 0                |                subrange 1                  |                subrange 2                  |                subrange 3                  | subrange 4
+        //   |        contains (2 * __n_sorted values)    |        contains (2 * __n_sorted values)    |        contains (2 * __n_sorted values)    |        contains (2 * __n_sorted values)    | contains the rest of data...  < Data parts
+        //   |----/----/----/----/----/----/----/----/----|----/----/----/----/----/----/----/----/----|----/----/----/----/----/----/----/----/----|----/----/----/----/----/----/----/----/----|----/---                       < Steps
+        //   ^              ^              ^              ^              ^              ^              ^              ^         ^    ^              ^              ^              ^              ^
+        //   |              |              |              |              |              |              |              |         |    |              |              |              |              |
+        // bd00           bd01           bd02           bd10           bd11           bd12           bd20           bd21        |  bd22           bd30           bd31           bd32           bd40                              < Base diagonals
+        //                  ^              ^              ^              ^              ^              ^              ^         |    ^              ^              ^              ^              ^
+        //  ---             0              1              2              3              4              5              6         |    7              8              9              10             11                              < Indexes in the base diagonal's SP storage
+        //   0    1    2    3    4    5    6    7    8    9    10   11   12   13   14  15   16    17   18   19   20   20   21   |    23   24   25   26   27   28   29   30   31   32   33   34   35    36                        < Linear IDs: __linear_id_in_steps_range
+        //   ^                                                                                                         |         |    |
+        //   |                                                                                                       __sp_left   |  __sp_right
+        //   |                                                                                                                   |
+        //   |                                                                                                       __linear_id_in_steps_range
+        //  We don't save the first diagonal into base diagonal's SP storage !!!
+
+        std::size_t __diagonal_idx = __linear_id_in_steps_range / __nd_range_params.steps_between_two_base_diags;
+
+        assert(__diagonal_idx < __nd_range_params.base_diag_count);
```

Comment:
Since we're in a kernel we should probably remove this (after debugging) until we have some utility to use `assert` with compilers we know support it in device code.

Reply:
Done